src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -211,32 +216,59 @@ private class ScopeTreeBuilder : BoundTreeWalkerWithStackGuardWithoutRecursionOn
                 /// </summary>
                 private readonly SmallDictionary<Symbol, Scope> _localToScope = new SmallDictionary<Symbol, Scope>();
 
-                private readonly Analysis _analysis;
+                private readonly MethodSymbol _topLevelMethod;
 
-                private ScopeTreeBuilder(Scope rootScope, Analysis analysis)
+                /// <summary>
+                /// If a local function is in the set, at some point in the code it is converted
+                /// to a delegate and should then not be optimized to a struct closure.
+                /// Also contains all lambdas (as they are converted to delegates implicitly).
+                /// </summary>
+                private readonly HashSet<MethodSymbol> _methodsConvertedToDelegates;
+                private readonly DiagnosticBag _diagnostics;
+
+                private bool _sawClosure = false;
```

Comment:
After instrumenting, the answer is yes: but it's tricky. The call sites look for nested functions before calling the LambdaRewriter, but it's actually possible in broken code for the LocalRewriter to find a lambda and remove it from the tree. This causes the LambdaRewriter to be called, but once we search the tree there are no lambdas remaining.

Catching this case doesn't seem worth the code complexity to me, so I'm just going to drop the check in the LambdaRewriter itself.